### ISSUE TITLE
Degenerate equalities of differential variables requiring specific values of parameters cannot be reduced

### ISSUE THREAD

**Issue by owiecc** [SciML/ModelingToolkit.jl#931]
I modified the [example](https://mtk.sciml.ai/stable/tutorials/acausal_components/) from documentation by having two inductors connected in series. This is a simplified MWE of the full code that I posted on [discourse](https://discourse.julialang.org/t/modelingtoolkits-structural-simplify-fails-on-li-cutset-simplification/57852):
```julia
using ModelingToolkit, DifferentialEquations

@parameters t L U
@variables i₁(t) i₂(t) u₁(t) u₂(t)

D = Differential(t)

eqs = [
        D(i₁) ~ u₁ / L
        D(i₂) ~ u₂ / L
        i₁ ~ i₂
        U ~ u₁ + u₂
      ]

ckt = ODESystem(eqs)

sys = structural_simplify(ckt)
```
`structural_simplify` throws an error:
```text
ERROR: LoadError: 

... Internal error in Tearing.jl: vs[1] = 0.

Stacktrace:
 [1] error(::String, ::Int64, ::String, ::Int64, ::String) at ./error.jl:42
 [2] initAlgebraicSystem(::ModelingToolkit.StructuralTransformations.TraverseDAG, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:116
 [3] tearEquations!(::ModelingToolkit.StructuralTransformations.TraverseDAG, ::Array{Array{Int64,1},1}, ::Array{Int64,1}, ::Array{Int64,1}; eSolvedFixed::Array{Int64,1}, vSolvedFixed::Array{Int64,1}, vTearFixed::Array{Int64,1}) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:320
 [4] tearEquations! at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:320 [inlined]
 [5] #28 at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:16 [inlined]
 [6] iterate at ./generator.jl:47 [inlined]
 [7] collect_to!(::Array{NTuple{4,Array{Int64,1}},1}, ::Base.Generator{Array{Array{Int64,1},1},ModelingToolkit.StructuralTransformations.var"#28#30"{SystemStructure,BipartiteGraph{Int64,Nothing},BipartiteGraph{Int64,Nothing},Array{Int64,1},Array{Int64,1}}}, ::Int64, ::Int64) at ./array.jl:732
 [8] collect_to_with_first! at ./array.jl:710 [inlined]
 [9] _collect(::Array{Array{Int64,1},1}, ::Base.Generator{Array{Array{Int64,1},1},ModelingToolkit.StructuralTransformations.var"#28#30"{SystemStructure,BipartiteGraph{Int64,Nothing},BipartiteGraph{Int64,Nothing},Array{Int64,1},Array{Int64,1}}}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./array.jl:704
 [10] collect_similar at ./array.jl:628 [inlined]
 [11] map at ./abstractarray.jl:2162 [inlined]
 [12] tear_graph(::ODESystem) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:11
 [13] tearing(::ODESystem; simplify::Bool) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:221
 [14] tearing(::ODESystem) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:221
 [15] structural_simplify(::ODESystem) at /Users/szymon/.julia/packages/ModelingToolkit/7KM6m/src/systems/abstractsystem.jl:526
 [16] top-level scope at /Users/szymon/Desktop/ModelingToolkit/test.jl:18
in expression starting at /Users/szymon/Desktop/ModelingToolkit/test.jl:18
```
`Project.toml` for reproducibility: 
```toml
[deps]
DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
```

**Comment by ChrisRackauckas** [comment]
Can you show the whole error message?

**Comment by baggepinnen** [comment]
This, or some similar error, usually happens to me unless I manually use `expand_derivatives` on the equations.

**Comment by owiecc** [comment]
@ChrisRackauckas I modified my question with the full stacktrace.

**Comment by YingboMa** [comment]
I think it's because alias elimination was not able to eliminate the degenerate states `i₁` or `i₂`
```julia
        D(i₁) ~ u₁ / L
        D(i₂) ~ u₂ / L
        i₁ ~ i₂
```
because they are differential variables. Is there a way to formulate your equations so that this doesn't happen?

**Comment by YingboMa** [comment]
```
        D(i₁) ~ u₁ / L
        D(i₂) ~ u₂ / L
        i₁ ~ i₂
        U ~ u₁ + u₂
```
implies
```
        D(i₁) ~ u₁ / L
        D(i₂) ~ (U - u₁) / L
        i₁ ~ i₂
```
and this system is degenerate structurally, because differential equations already uniquely specifies `i₁` and `i₂`. Also, we have no control over `U`, so we cannot possibly set `U` to be a state-varying parameter `2u₁`.

**Comment by YingboMa** [comment]
I don't think we can make this work. `U` is a state-varying parameter, which just means `U` should be a state. But then, you have 4 equations and 5 states.

**Comment by owiecc** [comment]
This kind of problem may randomly pop up when analyzing electrical circuits. Typically the issue is with two capacitors are connected in parallel (as this is something that is used very often). This can, of course, be mitigated by adding some helper elements in the circuit (a small capacitor in the node between inductors or a small inductor in the node between capacitors). 

This problem is solved e.g. here: Chen et al. “A practical regularization technique for modified nodal analysis in large-scale time-domain circuit simulation.” IEEE Transactions on Computer-Aided Design of Integrated Circuits and Systems 31.7 (2012): 1031-1040).

> These index-2 systems in most cases result from the presence of CV-loops and LI-cutsets in the topology [12]. Hence, the key to address the second bottleneck lies in reducing an index-2 circuit to its index-1 equivalent prior to the elimination process of (4) by breaking all CV- loops and LI-cutsets in the topology.

As far as I understand they effectively convert
```julia
D(i₁) ~ u₁ / L₁
D(i₂) ~ u₂ / L₂
i₁ ~ i₂
U ~ u₁ + u₂
```
into
```julia
D(i) ~ U / (L₁+L₂)
```
where these become observables
```julia
i₁ ~ i
i₂ ~ i
```
@YingboMa do you think this kind of transformation can be done?

**Comment by ChrisRackauckas** [comment]
I'll open as speculative. I think it's something that in theory could be done, but is probably much harder to do automatically than it is for a human to spot the transformation. I wouldn't rely on this happening soon, but keeping this open to capture real-world cases to then become a test suite for the eventual feature is probably a good idea.

**Comment by crlaugh** [comment]
Dymola can handle this case (though perhaps this set of equations is singular.  For the following Modelica code: 
```
model TestCircuit

  Modelica.Electrical.Analog.Basic.Inductor inductor1(L=1);
  Modelica.Electrical.Analog.Basic.Inductor inductor2(L=2);
  Modelica.Electrical.Analog.Basic.Ground ground;
  Modelica.Electrical.Analog.Sources.StepVoltage stepVoltage(V=1, startTime=1);
equation 
  connect(stepVoltage.p, inductor1.p);
  connect(inductor1.n, inductor2.p);
  connect(inductor2.n, ground.p);
  connect(ground.p, stepVoltage.n);
end TestCircuit;
```

which looks like this in graphical form: 
![Screenshot 2021-03-25 094001](https://user-images.githubusercontent.com/9616147/112482172-243ee800-8d4e-11eb-9bbf-a3359210f1a7.png)

the dsmodel.mof code is: 

```
// Translated Modelica model generated by Dymola from Modelica model 
//  SimpleCycleModels.Media.Testing.BuildingsMoistAir.TestCircuit

// -----------------------------------------------------------------------------

// Initial Section
  inductor2.n.v := 0;
  ground.p.v := 0.0;
  stepVoltage.n.v := 0.0;
  ground.p.i := 0.0;
// -----------------------------------------------------------------------------

// Bound Parameter Section
  stepVoltage.signalSource.height := stepVoltage.V;
  stepVoltage.signalSource.offset := stepVoltage.offset;
  stepVoltage.signalSource.startTime := stepVoltage.startTime;
// -----------------------------------------------------------------------------

// Initial Section
  ground.p.i := 0.0;

// -----------------------------------------------------------------------------

// Dynamics Section
  inductor1.p.v := stepVoltage.signalSource.offset+(if time < stepVoltage.signalSource.startTime
     then 0 else stepVoltage.signalSource.height);

  // Linear system of equations
  // Tag: simulation.linear[1]
    // Symbolic solution
      /*  Original equation
      inductor2.L*der(inductor1.i) = inductor1.n.v;
      */
      der(inductor1.i) := inductor1.p.v/(inductor2.L+inductor1.L);

    // Torn part
      inductor1.v := inductor1.L*der(inductor1.i);
      inductor1.n.v := inductor1.p.v-inductor1.v;
  // End of linear system of equations

// -----------------------------------------------------------------------------

// Eliminated alias variables
  der(inductor2.i) = der(inductor1.i);
  inductor1.n.i =  -inductor1.i;
  inductor1.p.i = inductor1.i;
  inductor2.i = inductor1.i;
  inductor2.n.i =  -inductor1.i;
  inductor2.p.i = inductor1.i;
  inductor2.p.v = inductor1.n.v;
  inductor2.v = inductor1.n.v;
  stepVoltage.i =  -inductor1.i;
  stepVoltage.n.i = inductor1.i;
  stepVoltage.p.i =  -inductor1.i;
  stepVoltage.p.v = inductor1.p.v;
  stepVoltage.signalSource.y = inductor1.p.v;
  stepVoltage.v = inductor1.p.v;
```

I believe that it uses the method of dummy derivatives to identify the variables to differentiate.

**Comment by YingboMa** [comment]
I will try out this system later this week. I think MTK can handle this system when it's formulated in component style.

**Comment by owiecc** [comment]
I posted the [code in component style](https://discourse.julialang.org/t/modelingtoolkits-structural-simplify-fails-on-li-cutset-simplification/57852) on discourse.

**Comment by YingboMa** [comment]
Alright. I think the component style model can be handled, but not your reduced model. It's better to also post the original system along with a reduced model, so that I can see what you are trying to do exactly.

**Comment by YingboMa** [comment]
@owiecc could you file another issue with the original system? I think handling your reduced system is a different issue, so it's better to keep that separate. I will work on fixed the original system, but this issue is still speculative.

**Comment by YingboMa** [comment]
I am pretty sure this is just the wrong model, so I am closing this.
